applicationset/generators/generator_spec_processor_test.go

```diff
@@ -0,0 +1,300 @@
+			directories:   []argoprojiov1alpha1.GitDirectoryGeneratorItem{{Path: "*"}},
+			repoApps:      []string{},
+			repoError:     fmt.Errorf("error"),
+			expected:      []map[string]string{},
+			expectedError: fmt.Errorf("error"),
+		},
+	}
+	argoCDServiceMock := argoCDServiceMock{mock: &mock.Mock{}}
+
+	argoCDServiceMock.mock.On("GetDirectories", mock.Anything, mock.Anything, mock.Anything).Return(cases[0].repoApps, cases[0].repoError)
+
+	var gitGenerator = NewGitGenerator(argoCDServiceMock)
+	return gitGenerator
+}
+
+func TestGetRelevantGenerators(t *testing.T) {
+
+	testGenerators := map[string]Generator{
+		"Clusters": getMockClusterGenerator(),
+		"Git":      getMockGitGenerator(),
+	}
+
+	testGenerators["Matrix"] = NewMatrixGenerator(testGenerators)
+	testGenerators["Merge"] = NewMergeGenerator(testGenerators)
+	testGenerators["List"] = NewListGenerator()
+
+	requestedGenerator := &argoprojiov1alpha1.ApplicationSetGenerator{
+		List: &argoprojiov1alpha1.ListGenerator{
+			Elements: []apiextensionsv1.JSON{{Raw: []byte(`{"cluster": "cluster","url": "url","values":{"foo":"bar"}}`)}},
+		}}
+
+	relevantGenerators := GetRelevantGenerators(requestedGenerator, testGenerators)
+	assert.Len(t, relevantGenerators, 1)
+	assert.IsType(t, &ListGenerator{}, relevantGenerators[0])
+
+	requestedGenerator = &argoprojiov1alpha1.ApplicationSetGenerator{
+		Clusters: &argoprojiov1alpha1.ClusterGenerator{
+			Selector: metav1.LabelSelector{},
+			Template: argoprojiov1alpha1.ApplicationSetTemplate{},
+			Values:   nil,
+		},
+	}
+
+	relevantGenerators = GetRelevantGenerators(requestedGenerator, testGenerators)
+	assert.Len(t, relevantGenerators, 1)
+	assert.IsType(t, &ClusterGenerator{}, relevantGenerators[0])
+
+	requestedGenerator = &argoprojiov1alpha1.ApplicationSetGenerator{
+		Git: &argoprojiov1alpha1.GitGenerator{
+			RepoURL:             "",
+			Directories:         nil,
+			Files:               nil,
+			Revision:            "",
+			RequeueAfterSeconds: nil,
+			Template:            argoprojiov1alpha1.ApplicationSetTemplate{},
+		},
+	}
+
+	relevantGenerators = GetRelevantGenerators(requestedGenerator, testGenerators)
+	assert.Len(t, relevantGenerators, 1)
+	assert.IsType(t, &GitGenerator{}, relevantGenerators[0])
+}
+
+func TestInterpolateGenerator(t *testing.T) {
+	requestedGenerator := &argoprojiov1alpha1.ApplicationSetGenerator{
+		Clusters: &argoprojiov1alpha1.ClusterGenerator{
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"argocd.argoproj.io/secret-type": "cluster",
+					"path-basename":                  "{{path.basename}}",
+					"path-zero":                      "{{path[0]}}",
+					"path-full":                      "{{path}}",
+				}},
+		},
+	}
+	//mockGitGenerator := getMockGitGenerator()
+	gitGeneratorParams := []map[string]string{
+		{"path": "p1/p2/app3", "path.basename": "app3", "path[0]": "p1", "path[1]": "p2", "path.basenameNormalized": "app3"},
+	}
```

Comment:
@KojoRising Regarding [this comment](https://github.com/argoproj/argo-cd/pull/9080#discussion_r853506892) and the screenshot with the arrows:

This `gitGeneratorParams` mock represents a `git` generator that yields a single result. My concern is, what happens when the `git` generator yields multiple results? _All_ of those results would be in `g0` over in `matrix.go` but `interpolateGenerator` is only interpolating values from the first result.

Reply:
@Lobstrosity 

Ahh, got it got it. Yes, you’re right, for multiple paths specified in a git generator, it wouldnt pick those up. 

I’ll update the code/test-cases to support this. Thank you for pointing it out!